### PR TITLE
Make `compCurBB` available for `fgMorphBlockReturn`.

### PR DESCRIPTION
**Pull request by sandreenko**
We were calling `tree = fgMorphCopyBlock(tree)`  in `fgMergeBlockReturn` and it could call `optAssertionGen(asg)` that would use `compCurBB` in several places like https://github.com/dotnet/runtime/blob/d6e3a0fcbb644c272c0dd2695c0053e62e2827c5/src/coreclr/src/jit/assertionprop.cpp#L2731

We have not seen a failure because we did not have struct returns and `tree->OperIsCopyBlkOp()` check was always false.

No diffs x64(crossgen, pmi, SPMI).

Easier to review by commits with whitespace changes disables:

04eccab3955: Extract `fgMergeBlockReturn`.
I wanted to call it `fgMorphBlockReturn`, but morph is already a very vague verb in the Jit, tried to use a more precise one.

315288b7712: Add a function header.

8347f1c5c72: Make `compCurBB` available for `fgMorphBlockReturn`.
When we generate an assignment we could need to create a new assertion, that requires `compCurBB` to be available.

c9ee223a537: Delete `INVALID_POINTER_VALUE`.
I would like to remove it because:
1) it was under debug only;
2) there were no null checks for `compHndBBtab` because it is a dependent variable
so there was no need to distinguish valid null pointer from a bad invalid pointer;
3) that was the only place where this mechanism was used.

1f60a915b67: Allow to CSE the merge return ASG.
I can't see a reason why it should not be allowed.
The issue with the previous version was that we did not actually know what we were marking: GT_ASG, GT_COMMA, something else? Was the idea to mark individual ASG under COMMA?

Failures are unrelated.

Preparation for #33225.

### CODE REVIEW

**Comment by sandreenko** [comment]
PTAL @dotnet/jit-contrib

File: src/coreclr/src/jit/morph.cpp [hunk 1]
```diff
@@ -15754,132 +15746,144 @@ void Compiler::fgMorphBlocks()
...
+//   It could end with a tail call or rejected tail call or monitor.exit or a GT_INTRINSIC.
+//   For now it is safe to explicitly check whether last stmt is GT_RETURN if genReturnLocal
+//   is BAD_VAR_NUM.
+//
+void Compiler::fgMergeBlockReturn(BasicBlock* block)
+{
+    assert((block->bbJumpKind == BBJ_RETURN) && ((block->bbFlags & BBF_HAS_JMP) == 0));
+    assert((genReturnBB != nullptr) && (genReturnBB != block));
+
+    // TODO: Need to characterize the last top level stmt of a block ending with BBJ_RETURN.
```
Comment (CarolEidt) [inline src/coreclr/src/jit/morph.cpp]: I realize this was pre-existing, but do you know what it means?

**Review by CarolEidt — APPROVED** [review]
LGTM

File: src/coreclr/src/jit/morph.cpp (hunk 1 above)
Comment (sandreenko) [reply src/coreclr/src/jit/morph.cpp]: I do not know, I guess the idea was to describe different patterns that we can see as the last STMT in a `BBJ_RETURN`, they are probably `GT_RETURN` tree, `GT_CALL (for a tail call)`, and anything else for a void return method?

**Pull request merged**